packages/sdks/typescript/src/index.ts

```diff
@@ -1,163 +1,212 @@
+      }
+      if (!event.data) {
+        onError?.(new Error(`Invalid data in server event:\n${event}`))
+        return
+      }
+      if (event.event === 'error') {
+        onError?.(new Error(event.data))
+        return
+      }
+
+      const json = this.parseJSON(event.data)
+      if (!json) {
+        const error = new Error(`Invalid JSON in server event:\n${event}`)
+        onError?.(error)
+        return
+      }
+
+      if (event.event === 'latitude-event') {
+        const messages = 'messages' in json ? (json.messages! as Message[]) : []
+
+        if (json.type === 'chain-error') {
+          const error = new Error((json.error as Error).message)
+          onError?.(error)
+
+          return
+        }
+
+        if (messages.length > 0) {
+          conversation.push(...messages)
+        }
+
+        if (json.type === 'chain-complete') {
+          uuid = json.uuid!
+          chainResponse = json.response!
+        }
+      }
 
-    const conversation: Message[] = []
-    let chainResponse: ChainCallResponse
+      onEvent?.({ event: event.event as StreamEventTypes, data: json })
+    }
+
+    const reader = stream.getReader()
     while (true) {
       const { done, value } = await reader.read()
-
       if (done) break
 
       const chunks = new TextDecoder('utf-8').decode(value).trim()
-      if (chunks?.startsWith('event: error')) {
-        onError?.(new Error(chunks))
-        break
-      }
-
-      chunks.split('\n').forEach((line) => {
-        if (!line.startsWith('data:')) return
-
-        const json = line.slice(6)
-        const chunk = this.parseJSON(json, onError)
-        if (!chunk) return
-
-        // FIXME: Magic variables. I think we need to share these constants (enums)
-        // in a new package. Now they are in `core` but I don't want to depend on core
-        // just for this. We need these enums in production not only as a dev dependency.
-        if (chunk.event === 'latitude-event') {
-          const messages = 'messages' in chunk.data ? chunk.data.messages! : []
-          if (messages.length > 0) {
-            // At the moment all message.content should be a string
-            // but in the future this could be an image or other type
-            // @ts-ignore
-            conversation.push(...messages)
-          }
-
-          if (chunk.data.type === 'chain-complete') {
-            chainResponse = chunk.data.response
-          }
+      for (const chunk of chunks.split('\n\n')) {
+        const event = this.parseEvent(chunk)
+        if (!event) {
+          onError?.(new Error(`Invalid SSE event: ${chunk}`))
+          return
+        }
+        const data = this.parseData(chunk)
+        if (data === null || data === undefined) {
+          onError?.(new Error(`Invalid data in server event:\n${chunk}`))
+          return
         }
 
-        onMessage?.(chunk)
-      })
+        consumeEvent({ event, data })
+      }
     }
 
+    if (!uuid || !chainResponse) return
```

Comment:
it means the streaming failed at some point. Should never be true but better safe than sorry.